adsdata/nonbib.py

```diff
@@ -335,6 +343,30 @@ def count_lines(self, file):
             on {0}.canonical.bibcode = {1}.canonical.bibcode \
             where {1}.canonical.bibcode IS NULL;'
 
+    # resolver
+    include_new_resolver_bibcodes_sql = \
+        'insert into {0}.ChangedRowsM (bibcode) \
+            select distinct on (bibcode) bibcode from \
+               (select {0}.datalinks.bibcode from {0}.datalinks left join {1}.datalinks \
+                on {0}.datalinks.bibcode = {1}.datalinks.bibcode \
+                and {0}.datalinks.link_type = {1}.datalinks.link_type \
+                and {0}.datalinks.link_sub_type = {1}.datalinks.link_sub_type \
+                where {0}.datalinks.url != {1}.datalinks.url \
+                or {0}.datalinks.title != {1}.datalinks.title \
+                or {0}.datalinks.item_count != {1}.datalinks.item_count \
+                or {1}.datalinks.bibcode IS NULL) as datalinks;'
```

Comment:
To be safe and make sure that we do not insert already existing bibcodes into `ChangedRowsM`, I suggest we add this to the end:

```
... as datalinks where not exists ( select 'x' from {0}.ChangedRowsM where {0}.ChangedRowsM.bibcode = bibcode);
```

I tested it and it does not make the query slower, while it protect us from inserting bibcodes that already existed. This could happen for instance if the bibcode had a change in the number of reads (so it was inserted before) and a change in a URL (so we would try to insert it again).

Reply:
that is what distinct does actually. "With DISTINCT ON , You tell PostgreSQL to return a single row for each distinct group defined by the ON clause. Which row in that group is returned is specified with the ORDER BY clause. That's it!" so distinct makes sure if multiple link_type and link_sub_types have been updated the bibcode is returned only once.